### DISCUSSION TITLE
Is the memory usage dependent on the number of objects within the input file?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#19144] [Q&A General] [answered]
Hi guys,

I'm wondering if the memory usage is dependent on the number of objects within the input file? 

I was testing my hydro simulation on a large and complex mesh (5mill elements/1.3mill nodes).
when I run the Steady-state test, I could run it with 20+ cores mpi (with 256gb RAM workstation), but then when I tried to run it with transient mode, with 120+ BCs objects and 120+ Control objects (the input file is shown below), the memory usage blew up and I couldn't run it with more than 10 cores mpi.

```
BC block
     [./drain_DRAINSRF_CUMSRFSET_Y2020_M07]
     type = PorousFlowPiecewiseLinearSinkTC
     variable = porepressure
     boundary = 'DRAINSRF_CUMSRFSET_Y2020_M07'
     pt_vals = '0 1e9'
     multipliers = '0 1e9'
     flux_function = 1e5
     v = min
      use_mobility = true
      fluid_phase = 0
     [../]
     [./drain_DRAINSRF_CUMSRFSET_Y2020_M11]
     type = PorousFlowPiecewiseLinearSinkTC
     variable = porepressure
     boundary = 'DRAINSRF_CUMSRFSET_Y2020_M11'
     pt_vals = '0 1e9'
     multipliers = '0 1e9'
     flux_function = 1e5
     v = min
      use_mobility = true
      fluid_phase = 0
     [../]
Control block
    [water_grad_drain_DRAINSRF_CUMSRFSET_Y2020_M07]
    type = ConditionalFunctionEnableControl
    enable_objects = 'BCs::drain_DRAINSRF_CUMSRFSET_Y2020_M07'
    conditional_function = 'if(t=5,1,0)'
    execute_on = 'initial timestep_begin'
    []
    [water_grad_drain_DRAINSRF_CUMSRFSET_Y2020_M11]
    type = ConditionalFunctionEnableControl
    enable_objects = 'BCs::drain_DRAINSRF_CUMSRFSET_Y2020_M11'
    conditional_function = 'if(t=6,1,0)'
    execute_on = 'initial timestep_begin'
    []
```

From #18930 I remember @GiudGiud mentioned that:

> Not all systems in MOOSE are scalable, and MOOSE generally should not scale well with the input file length because every object is created by every process. Then based on the mesh distribution only some of the objects remain active locally.

So if that's the case, how should I work around this?

Is it possible to modify the name/value within the block at each time step? 

for example

```
t =1 
    boundary = 'DRAINSRF_CUMSRFSET_Y2020_M11'
    enable_objects = 'BCs::drain_DRAINSRF_CUMSRFSET_Y2020_M07'
    conditional_function = 'if(t=5,1,0)'
t = 2 
    boundary = 'DRAINSRF_CUMSRFSET_Y2020_M11'
    enable_objects = 'BCs::drain_DRAINSRF_CUMSRFSET_Y2020_M11'
    conditional_function = 'if(t=6,1,0)'
```

I guess this could be done via `Action`, but would this solve the problem?

This is because at some point we would need to work with more than 500+ control/bc object, overcoming this issue at early stage would be nice.

Kind regards,
Traiwit

**Comment by YaqiWang** [comment, marked as answer]
Every processor holds its own set of kernels/bcs, etc. I think on average a kernel or a bc cost about 200kB memory (this number may not be very accurate). If we have 500 objects, the memory used for storing these objects on each processor would be 500*200kB=100MB. This memory is not scalable and an overhead. We met the situation in radiation transport and we ended up re-implement the kernels with array variable/kernel, which significantly reduced the number of kernels. I think MOOSE block restriction can introduce significant memory overhead as well if you have lots of mesh blocks (subdomains). In such cases, you may want to consider minimizing the number of mesh blocks by merging some of them without affecting your calculation if possible. Same applies to mesh boundaries. Unfortunately I cannot provide exact numbers to support my solutions here but that should give you a rough idea. Not sure if the latest live print capability can show you where the memory hot spots are. Another useful parameter is `Debug/show_actions` that can print the memory usage at the beginning of each action. Hope these help.

**Reply by GiudGiud** [reply]
I think the show_actions log will be important here. 25 gb per rank seems very excessive.

**Reply by Traiwit** [reply]
Hi guys, I'm sending part that the memory is blowing up below, i think it mainly comes from the 400,000 constraints blocks that I got in the action that i created @GiudGiud

a quick note that, I think this mesh I'm using has 1500 subdomains 

```
[DBG][ACT] TASK ( execute_mesh_generators) TYPE (           ExecuteMeshGenerators) NAME (                ) Memory usage 167MB
..............................................
[DBG][ACT] TASK (       recover_meta_data) TYPE (      SetupRecoverFileBaseAction) NAME (                ) Memory usage 4796MB
[DBG][ACT] TASK (           set_mesh_base) TYPE (                 SetupMeshAction) NAME (            Mesh) Memory usage 4796MB
[DBG][ACT] TASK (     attach_geometric_rm) TYPE (          AddRelationshipManager) NAME (                ) Memory usage 4796MB
[DBG][ACT] TASK (               init_mesh) TYPE (                 SetupMeshAction) NAME (            Mesh) Memory usage 4796MB
[DBG][ACT] TASK (            prepare_mesh) TYPE (         SetupMeshCompleteAction) NAME (            Mesh) Memory usage 4796MB
[DBG][ACT] TASK (                  add_ic) TYPE (       AddInitialConditionAction) NAME (    porepressure) Memory usage 5269MB
[DBG][ACT] TASK (          add_constraint) TYPE (                           MPCbe) NAME (           MPCbe) Memory usage 5269MB
.
.
.
Still Setting Up............
[DBG][ACT] TASK (     add_preconditioning) TYPE (       SetupPreconditionerAction) NAME (             SMP) Memory usage 19236MB
[DBG][ACT] TASK (           ready_to_init) TYPE (                     EmptyAction) NAME (                ) Memory usage 19236MB
[DBG][ACT] TASK (           setup_dampers) TYPE (              SetupDampersAction) NAME (                ) Memory usage 19236MB
[DBG][ACT] TASK (    setup_residual_debug) TYPE (        SetupResidualDebugAction) NAME (           Debug) Memory usage 19236MB
[DBG][ACT] TASK (         copy_nodal_vars) TYPE (             CopyNodalVarsAction) NAME (    porepressure) Memory usage 19236MB
[DBG][ACT] TASK (          add_aux_kernel) TYPE (            MaterialOutputAction) NAME (                ) Memory usage 19545MB
[DBG][ACT] TASK (  coupling_functor_check) TYPE (      CouplingFunctorCheckAction) NAME (coupling_functor_check) Memory usage 19545MB
[DBG][ACT] TASK (            init_problem) TYPE (               InitProblemAction) NAME (                ) Memory usage 19545MB

  Initializing
    Updating Because Mesh Changed
      Updating Mesh..
        Building Node List                                                               [ 10.50 s] [    1 MB]
        Finished Building Boundary Elements List                                         [  9.16 s] [    0 MB]
      Still Updating Mesh.                                                               [ 53.11 s] [    1 MB]
    Finished Updating Because Mesh Changed                                               [ 53.52 s] [  127 MB]
      Building Node To Elem Map                                                          [  8.43 s] [  304 MB]
    Still Initializing Equation Systems.....                                             [ 52.58 s] [ 3800 MB]
[DBG][ACT] TASK (             add_control) TYPE (                AddControlAction) NAME (water_grad_drain_DRAINSRF_CUMSRFSET_Y2012) Memory usage 24030MB
[DBG][ACT] TASK (             add_control) TYPE (                AddControlAction) NAME (water_grad_drain_DRAINSRF_CUMSRFSET_Y2015) Memory usage 24030MB
[DBG][ACT] TASK (             add_control) TYPE (                AddControlAction) NAME (water_grad_drain_DRAINSRF_CUMSRFSET_Y2019) Memory usage 24031MB
[DBG][ACT] TASK (             add_control) TYPE (                AddControlAction) NAME (water_grad_drain_DRAINSRF_CUMSRFSET_Y2020_M07) Memory usage 24031MB
```

so mesh generator 167 > 4800
constraints 5250 > 19000
InitProblemAction 19000 > 24000

Yeah @GiudGiud looks like it's ~25GB per rank which is mainly from constraints (guess nothing I can do about this unless there is a way to optimise it)

**Reply by GiudGiud** [reply]
So 5GB roughly for the mesh, and 5GB for the equation system. Then 14GB for the constraints. 
Except the first two can be split among all processes, while the last one currently is not. 

Reworking the constraint to work with a single constraint dealing with large lists of nodes is an option, but it will require modifying the routines handling the contribution to residuals & jacobian. 

I think the easier option here would be to modify your constraints Action to make it only add constraints that are local to each process. So pass a mesh argument to this action and look for the nodes in the local mesh (mesh.get_active_node_ptr_range or something like that) and only add if the constraint is local or partially local to each process.

**Reply by Traiwit** [reply]
> Except the first two can be split among all processes, while the last one currently is not

what do you mean by 'split among all processes', I thought this 5 5 14 gbs are per core already.

> I think the easier option here would be to modify your constraints Action to make it only add constraints that are local to each process. So pass a mesh argument to this action and look for the nodes in the local mesh (mesh.get_active_node_ptr_range or something like that) and only add if the constraint is local or partially local to each process.

@GiudGiud I reckon this is the ideal solution, if we do this, let's say we have 10 local meshes, then it will only be 1.4gbs per core? Is there any example of doing this? 

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
not that I know of. @rwcarlsen do you know anyone doing that?
Roughly. It wont be perfectly split since some constraints will be split between two domains. But it should be pretty close

**Reply by GiudGiud** [reply]
We've chatted about it and one thing you ll have to be mindful of is not to repartition during the simulation. There may be other difficulties if you end up with a process with 0 constraint, so make sure to have at least one on each process, even if it s inactive (because it uses nodes that are not on the process' mesh)
